controller/api/destination/server.go

```diff
@@ -364,17 +364,44 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 	return nil
 }
 
-func (s *server) createAddress(pod *corev1.Pod, port uint32) (watcher.Address, error) {
+// getPortForPod returns the port that a `pod` is listening on.
+//
+// If the `host` address given is the HostIP, use the HostPort mapping for the `pod`
+// to determine the port to return.
+func (s *server) getPortForPod(pod *corev1.Pod, host string, port uint32) (uint32, error) {
+	if host == pod.Status.PodIP {
+		return port, nil
+	}
+
+	if host == pod.Status.HostIP {
+		for _, container := range pod.Spec.Containers {
+			for _, containerPort := range container.Ports {
+				if uint32(containerPort.HostPort) == port {
+					return uint32(containerPort.ContainerPort), nil
+				}
+			}
+		}
+	}
+
+	return 0, fmt.Errorf("unable to find container port as host (%s) matches neither PodIP nor HostIP (%s)", host, pod)
```

Comment:
Do we really need to return an error in this case? It seems like we could log a warning when the target IP is unknown, but we can proceed using the original port, probably.